### PR TITLE
Add hash-mode 3730 - Dahua NVR/DVR - md5($salt1.strtoupper(md5($salt2…

### PR DESCRIPTION
**Pull request by davidrozen76**
….$pass)))

Pure kernel on attack mode 0 doesn't crack when only a single hash is present in the hashfile (otherwise ok) - self-test fails as well.

In accordance to issue #3468

### CODE REVIEW

File: OpenCL/m03730_a0-pure.cl [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+  for (u32 i = 0, idx = 0; i < salt2_len; i += 4, idx += 1)
+  {
+    s2[idx] = esalt_bufs[DIGESTS_OFFSET_HOST].salt2_buf[idx];
+  }
+
+  md5_ctx_t ctx0;
+
+  md5_init (&ctx0);
+
+  md5_update_global (&ctx0, esalt_bufs[DIGESTS_OFFSET_HOST].salt1_buf, esalt_bufs[DIGESTS_OFFSET_HOST].salt1_len);
```
Comment (piwvvo) [inline OpenCL/m03730_a0-pure.cl]: It seems that single hash kernel doesn't work correctly because you used salt1_buf here instead of salt2_buf

**Comment by jsteube** [comment]
Please provide unit-test in tools/test_modules for this plugin. Shouldn't be too hard to realize there's plenty of other modules to copy the code from and then just adapt. But much more important is to remove the changes to inc_common.h. If you have a fast salt with esalt, there's also plenty of other modules than can be used as template.

**Comment by davidrozen76** [comment]
The unit-test fails on optimized kernels when salt length exceeds 16 characters.

Changes are to be made to support long salt lengths (a must for this hash-mode).

Any help will be appreciated :D

**Comment by jsteube** [comment]
Any handling of data larger the block size of the underlaying primitive in question requires branching. This takes a lot of effort. Of course I can't merge an unfinished plugin. If you want me to merge this, please remove the optimized kernels, because the pure kernels should work. We can later add the optimized kernels.

**Comment by jsteube** [comment]
Very cool, many thanks!

**Pull request merged**